draft-ietf-jsonpath-base.md

```diff
@@ -1521,13 +1520,15 @@ A function expression is well-typed if all of the following are true:
   to have a result type of `LogicalType`, `SingleNodeType`, or `NodesType`.
 * If it occurs directly as a `comparable` in a comparison, the
   function is declared to have a result type of `ValueType` or `SingleNodeType`.
-* Otherwise, it is occurring as an argument to a further function
-  expression, and the following rules for function arguments apply to
+* Otherwise, it occurs as an argument to a further function
+  expression and the following rules for function arguments apply to
   its declared result type.
-* Each argument of the function can be used for the declared type of the corresponding declared
+* Each argument of the function conforms to the declared type of the corresponding declared
   parameter according to one of the following rules:
-   * The argument is a function expression with declared result type that is the same as,
-     or a subtype of, the declared type of the parameter.
+   * The argument is a function expression with declared result type that is the same as
+     the declared type of the parameter.
+   * The argument is a function expression with declared result type `SingleNodeType` and
+     the declared type of the parameter is `NodesType`.
```

Comment:
> `SingleNodeType` should be valid for both.

@gregsdennis That sounds plausible, but wouldn't we also need to modify the note below Table 13:

> An instance of `SingleNodeType` is well-typed wherever an instance of `NodesType` is well-typed.

to read:

> An instance of `SingleNodeType` is well-typed wherever an instance of `NodesType` or `ValueType` is well-typed.

?

Reply:
On reflection, it may be better to avoid this implicit type conversion in the function side of the spec and use a conversion function instead (as part of adding various such functions, see https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/412#discussion_r1121407915).

Reply:
> wouldn't we also need to modify the note below Table 13

Yes, that would need to update as well.

Reply:
> On reflection, it may be better to avoid this implicit type conversion in the function side of the spec and use a conversion function instead (as part of adding various such functions, see https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/412#discussion_r1121407915).

What implicit type conversion do you still see?  You've explicitly specified away all of the conversions.

Reply:
`SingleNodeType` to `ValueType` (in the case where a `SingleNodeType` function expression is used as an argument for a `ValueType` parameter). This isn't covered by the non-function side of the spec.

Reply:
Maybe that just needs an explicit statement then.

> Singular Paths and functions which return `SingleNodeType` are valid as arguments for function parameters expressed as `ValueType`.  In these cases, the function parameter receives the value of the single node.

or something.

Reply:
That would certainly work, but I'm not sure the it makes the spec particularly clear. I'm currently favouring explicit type conversions in the function world and implicit conversions only at the boundary between the function world and the rest of the spec.

Reply:
> I'm currently favouring explicit type conversions in the function world and implicit conversions only at the boundary between the function world and the rest of the spec.

This is an inconsistency, though.  There's no reason to treat function syntax any differently than the rest of the expression.  This is what I'm hung up on that both of you keep pushing.

Functions are part of the overall expression grammar.  The expression as a whole needs certain rules, and every part of the expression should follow those rules, including inside and outside of functions.

Reply:
> > I'm currently favouring explicit type conversions in the function world and implicit conversions only at the boundary between the function world and the rest of the spec.
> 
> This is an inconsistency, though.

> There's no reason to treat function syntax any differently than the rest of the expression.

I haven't explored this option because I suspect it would be less understandable than the alternatives we are currently exploring. Since you have a different intuition, please feel free to submit a PR to explore that part of the design space and then we can compare it with the alternatives.

>This is what I'm hung up on that both of you keep pushing.
> 
> Functions are part of the overall expression grammar. The expression as a whole needs certain rules, and every part of the expression should follow those rules, including inside and outside of functions.

Please correct me if I've misunderstood, but I think you are talking about referential transparency (the ability to use expressions with the same value interchangeably) and I agree that would be a valuable property. I'm just not sure how to achieve it in general.

> 
> I think those rules are fine being expressed as conversion; you think they need to be expressed more concretely. I'm fine with either so long as they apply everywhere.

That suggests two possible options in the design space. One is to roll out implicit conversions consistently everywhere which I suspect will not be easy for readers of the spec to understand (see my comment above). The other is to roll out the use of conversion functions everywhere, but I think that would break a lot of existing queries.

Reply:
I don't think there is any problem with the function expression space to be different (which Greg calls "inconsistent") from the grammar-based space.  You can do more with function expressions, so it is natural that you have to be more specific when using them.  Everything else being equal, you would want to avoid exposing these differences, but everything else is not equal.

Reply:
> I don't think there is any problem with the function expression space to be different

We're defining an expression grammar.  Rules that we apply in one part need to apply to the whole grammar.

Reply:
To drive this home, we're describing `NodesType` as an abstraction of `filter-path`, which is a Path.  This implies that Paths have a type, and that type is `NodesType`.